lib/logical_solver/presolve.cpp

```diff
@@ -0,0 +1,889 @@
+#include "presolve.hpp"
+#include <iostream>
+
+namespace logical_solver{
+
+  Presolve::Presolve(
+    const std::vector<std::vector<int>> problem_matrix, 
+    const std::vector<int> lower_bounds, 
+    const std::vector<int> upper_bounds,
+    const int inequalities_count,
+    const int equalities_count,
+    const bool solve_MIP
+  ) {
+    // Infitnity - eventually to be defined in types.
+    kInfinity = std::numeric_limits<int>::max(); 
+
+    // Define problem
+    problem_matrix_ = problem_matrix;
+
+    // Problem type
+    solve_MIP_ = solve_MIP;
+
+    // Get problem attributes
+    variables_count_ = problem_matrix[0].size();
+    constraints_count_ = problem_matrix.size();
+    inequalities_count_ = inequalities_count;
+    equalities_count_ = equalities_count;
+
+    // Set up constraints lower and upper bounds as 
+    // vector of doubles.
+    for (int i=0; i < constraints_count_; ++i) {
+      lower_bounds_.push_back((double)lower_bounds.at(i));
+      upper_bounds_.push_back((double)upper_bounds.at(i));
+    }
+
+    // Set up active rows and columns arrays as well
+    // as implied bounds using problem size and feasible
+    // solution vector.
+    presolve_active_rows_.resize(constraints_count_, true);
+    presolve_active_columns_.resize(variables_count_, true);
+    inequality_singletons_ = {};
+
+    implied_lower_bounds_.resize(variables_count_, -kInfinity);
+    implied_upper_bounds_.resize(variables_count_, kInfinity);
+    feasible_solution.resize(variables_count_, -999);
+
+    reduced_to_empty = false;
+    infeasible = false;
+    infeasible_by_PR = false;
+    unsatisfied_constraints = false;
+
+    presolve_active_rows_count_ = constraints_count_;
+    presolve_active_columns_count = variables_count_;
+  };
+
+  void Presolve::getRowsAndColsNonZeros() {
+    rows_non_zero_variables_.clear();
+    cols_non_zeros_indices_.clear();
+    rows_non_zero_variables_.resize(constraints_count_, {});
+    cols_non_zeros_indices_.resize(variables_count_, {});
+
+    for (int i = 0; i < constraints_count_; i++) {
+      if (presolve_active_rows_.at(i)) {
+
+        for (int j = 0; j < variables_count_; ++j) {
+          if (presolve_active_columns_.at(j)) {
+
+            if (problem_matrix_.at(i).at(j) != 0) {
+              rows_non_zero_variables_.at(i).push_back(j);
+              cols_non_zeros_indices_.at(j).push_back(i);
+            }
+          }
+        }
+      }
+    }
+  };
+
+  double Presolve::getFeasibleValueCalculationBound(int row_index) {
+    double lower_bound = lower_bounds_.at(row_index);
+    double upper_bound = upper_bounds_.at(row_index);
+
+    if (lower_bound != -kInfinity) {
+      return lower_bound;
+    }
+    return upper_bound;
+  }
+
+  int Presolve::getVariableFeasibleValueMIP(
+    int row_index, int col_index, 
```

Comment:
If it's an index (and hence should only ever be strictly positive), maybe using an unsigned integer such as `std::uint32_t` may be useful to avoid undefined behavior and would be better practice. Not essential tho.

Reply:
also these function arguments should be `const`.

Reply:
Actually check everywhere for const! Anything that can be const should be!

Reply:
Added all the consts but will leave it as an int normal int.